From: Julien Cristau <jcristau@debian.org>
Date: Tue, 19 Aug 2008 22:43:47 +0200
Subject: Fix compiler warnings

* glibc wants _XOPEN_SOURCE_EXTENDED for vfork(), and
  _BSD_SOURCE for strcasecmp, so let's define _GNU_SOURCE and get them all.
* greeter/Login.c: gcc gets confused about lastspace being initialized
* greeter/greet.c: add const modifier to pam_message struct
* Also cast away a bunch of signed vs unsigned char* warnings.

diff --git a/configure.ac b/configure.ac
index 329c8e8..ff9cfb5 100644
--- a/configure.ac
+++ b/configure.ac
@@ -123,7 +123,7 @@ fi
 
 case $host_os in
     linux*|gnu*|*-gnu)
-        OS_CFLAGS="-D_XOPEN_SOURCE"
+        OS_CFLAGS="-D_GNU_SOURCE"
 	SU="su"
         ;;
     darwin*)
diff --git a/genauth.c b/genauth.c
index 0b54b8d..245a63f 100644
--- a/genauth.c
+++ b/genauth.c
@@ -377,7 +377,7 @@ InitXdmcpWrapper (void)
 #endif    
     /*  Try some pseudo-random number genrator daemon next */
     if (prngdSocket != NULL || prngdPort != 0) {
-	    if (get_prngd_bytes(tmpkey, sizeof(tmpkey), prngdPort, 
+	    if (get_prngd_bytes((char *)tmpkey, sizeof(tmpkey), prngdPort, 
 		    prngdSocket) == 0) {
 		    tmpkey[0] = 0;
 		    _XdmcpWrapperToOddParity(tmpkey, key);
diff --git a/greeter/Login.c b/greeter/Login.c
index a156762..741724d 100644
--- a/greeter/Login.c
+++ b/greeter/Login.c
@@ -73,6 +73,7 @@ from The Open Group.
 # include <stdio.h>
 # include <ctype.h>
 # include <time.h>
+# include <string.h>
 
 # include "dm.h"
 # include "dm_error.h"
@@ -602,7 +603,7 @@ RedrawFail (LoginWidget w)
 	    char *tempCopy = strdup(w->login.fail);
 	    if (tempCopy != NULL) {
 		char *start, *next;
-		char lastspace;
+		char lastspace = '\0';
 
 		y = PROMPT_Y(w,LAST_PROMPT) + (2 * PROMPT_Y_INC(w));
 		
diff --git a/greeter/greet.c b/greeter/greet.c
index 0aabec7..27732ef 100644
--- a/greeter/greet.c
+++ b/greeter/greet.c
@@ -683,7 +683,7 @@ static int pamconv(int num_msg,
 	    "PAM_PROMPT_ECHO_OFF", "PAM_PROMPT_ECHO_ON",
 	    "PAM_ERROR_MSG", "PAM_TEXT_INFO" } ;
     
-    struct pam_message      *m;
+    const struct pam_message      *m;
     struct pam_response     *r;
 
     struct myconv_data	    *d = (struct myconv_data *) appdata_ptr;
diff --git a/xdmauth.c b/xdmauth.c
index b54d5ce..fc47821 100644
--- a/xdmauth.c
+++ b/xdmauth.c
@@ -182,7 +182,9 @@ XdmGetXdmcpAuth (struct protoDisplay *pdpy,
     XdmPrintDataHex ("Accept packet auth", xdmcpauth->data, xdmcpauth->data_length);
     XdmPrintDataHex ("Auth file auth", fileauth->data, fileauth->data_length);
     /* encrypt the session key for its trip back to the server */
-    XdmcpWrap (xdmcpauth->data, (unsigned char *)&pdpy->key, xdmcpauth->data, 8);
+    XdmcpWrap ((unsigned char *)xdmcpauth->data,
+               (unsigned char *)&pdpy->key,
+               (unsigned char *)xdmcpauth->data, 8);
     pdpy->fileAuthorization = fileauth;
     pdpy->xdmcpAuthorization = xdmcpauth;
 }
